context/context.go

```diff
@@ -1,60 +1,69 @@
-// Package context is the runtime context of GlusterD
-//
-// Any package that needs access to the GlusterD runtime context just needs to
-// import this package.
 package context
 
 import (
-	"sync"
-
-	"github.com/gluster/glusterd2/rest"
-	"github.com/gluster/glusterd2/transaction"
-
 	log "github.com/Sirupsen/logrus"
-	"github.com/pborman/uuid"
 )
 
-// Various version constants that will be used by GD2
-const (
-	MaxOpVersion    = 40000
-	APIVersion      = 1
-	GlusterdVersion = "4.0-dev"
-)
+// Context is used to carry contextual information across the lifetime of a request or a transaction.
+type Context struct {
+	parent *Context
+	data   map[string]interface{}
 
-// Any object that is a part of the GlusterD context and needs to be available
-// to other packages should be declared here as exported global variables
-var (
-	MyUUID    uuid.UUID
-	Rest      *rest.GDRest
-	TxnFw     *transaction.GDTxnFw
-	OpVersion int
-)
+	Log *log.Entry // Functions which are given this context must use this logger to log their data.
+}
 
-var (
-	initOnce sync.Once
-)
+// NewEmptyContext returns a new empty Context with no parent, no associated data and the default logger.
+func NewEmptyContext() *Context {
+	return &Context{
+		data: make(map[string]interface{}),
+		Log:  log.NewEntry(log.StandardLogger()), //empty logging context
+	}
+}
+
+// NewLoggingContext returns a new context with the logger set to log given fields
+func NewLoggingContext(fields log.Fields) *Context {
+	c := NewEmptyContext()
+	c.Log = c.Log.WithFields(fields)
 
-func initOpVersion() {
-	//TODO : Need cluster awareness and then decide the op-version
-	OpVersion = MaxOpVersion
+	return c
 }
 
-func doInit() {
-	log.Debug("Initializing GlusterD context")
+// NewContext returns a new empty context with given parent
+func (c *Context) NewContext() *Context {
+	return &Context{
+		parent: c,
+		data:   make(map[string]interface{}),
+		Log:    c.Log,
+	}
+}
 
-	initLocalStateDir()
+// NewLoggingContext returns a new context with the logger set to log given fields in addition to the parents logging fields
+func (c *Context) NewLoggingContext(fields log.Fields) *Context {
```

Comment:
NewLoggingContext () is already defined at line 24. Same function name is used here as well. What's the difference?